### PR TITLE
Unhandled intercept v2 (implement WIP-0018)

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Implement [WIP-0018](https://github.com/witnet/WIPs/blob/master/wip-0018.md)

TODO: the signaling bit is not assigned yet, but we may consider to merge wips 0017-0019.

### CODE REVIEW

**Comment by tmpolaczyk** [comment]
The latest commit changes this draft to apply only in the tally stage. Explanation below:

So there are two kinds of `UnhandledIntercept` errors, depending on where they are created:

* Created in tally

* Created in retrieval/aggregation

### Created in tally

WIP-0018 focuses on those that are created in the tally stage, because they are the consensus-critial ones. As an example, see data request `63c1ff0bacd24676bd71fa2bb984e8faee22cd8c73b14b83c57149a792dc315c`. In that request every node reports a different string, and the mode operator returns an error because there is no "most common string". This is the tally using the old validation rules:

```
Tally: RadonTypes::RadonError(UnhandledIntercept { inner: None, message: Some("inner: ModeTie { values: RadonArray { value: [String(RadonString { value: \"149.28.82.190\" }), String(RadonString { value: \"138.201.132.240\" }), String(RadonString { value: \"199.192.30.223\" }), String(RadonString { value: \"159.89.7.193\" }), String(RadonString { value: \"138.201.65.62\" }), String(RadonString { value: \"195.201.61.247\" }), String(RadonString { value: \"149.28.82.190\" }), String(RadonString { value: \"138.201.132.40\" }), String(RadonString { value: \"136.243.93.244\" }), String(RadonString { value: \"88.99.68.56\" }), String(RadonString { value: \"88.198.7.148\" }), String(RadonString { value: \"178.63.54.115\" }), String(RadonString { value: \"168.119.172.143\" }), String(RadonString { value: \"213.239.234.132\" }), String(RadonString { value: \"63.250.47.141\" }), String(RadonString { value: \"207.154.238.90\" }), String(RadonString { value: \"168.119.15.50\" }), String(RadonString { value: \"136.244.77.174\" }), String(RadonString { value: \"94.130.223.45\" }), String(RadonString { value: \"136.243.109.250\" }), String(RadonString { value: \"168.119.119.170\" }), String(RadonString { value: \"188.40.94.103\" }), String(RadonString { value: \"46.4.75.56\" }), String(RadonString { value: \"138.201.46.114\" }), String(RadonString { value: \"168.119.119.170\" }), String(RadonString { value: \"138.201.132.40\" }), String(RadonString { value: \"95.111.244.179\" }), String(RadonString { value: \"135.181.60.181\" }), String(RadonString { value: \"68.183.39.75\" }), String(RadonString { value: \"159.69.75.174\" }), String(RadonString { value: \"159.69.72.188\" }), String(RadonString { value: \"78.46.40.149\" }), String(RadonString { value: \"176.9.66.252\" }), String(RadonString { value: \"176.9.120.66\" }), String(RadonString { value: \"148.251.128.26\" }), String(RadonString { value: \"78.46.91.66\" }), String(RadonString { value: \"135.181.60.189\" }), String(RadonString { value: \"142.93.250.57\" }), String(RadonString { value: \"49.12.73.196\" }), String(RadonString { value: \"178.63.70.230\" }), String(RadonString { value: \"159.89.7.193\" }), String(RadonString { value: \"176.9.3.164\" }), String(RadonString { value: \"88.99.68.109\" }), String(RadonString { value: \"116.202.218.95\" }), String(RadonString { value: \"168.119.172.143\" }), String(RadonString { value: \"85.114.132.71\" }), String(RadonString { value: \"167.99.246.148\" }), String(RadonString { value: \"168.119.119.170\" }), String(RadonString { value: \"138.201.66.13\" }), String(RadonString { value: \"136.243.94.171\" }), String(RadonString { value: \"88.99.68.109\" }), String(RadonString { value: \"46.4.88.189\" }), String(RadonString { value: \"135.181.60.183\" }), String(RadonString { value: \"138.201.83.20\" }), String(RadonString { value: \"149.7.16.186\" }), String(RadonString { value: \"116.203.225.48\" }), String(RadonString { value: \"167.172.39.205\" }), String(RadonString { value: \"159.69.75.174\" }), String(RadonString { value: \"46.4.217.205\" }), String(RadonString { value: \"135.181.56.50\" }), String(RadonString { value: \"168.119.172.143\" }), String(RadonString { value: \"136.244.77.174\" }), String(RadonString { value: \"116.203.46.135\" }), String(RadonString { value: \"85.114.132.62\" }), String(RadonString { value: \"148.251.128.18\" }), String(RadonString { value: \"148.251.152.217\" }), String(RadonString { value: \"136.243.64.186\" }), String(RadonString { value: \"159.89.7.193\" }), String(RadonString { value: \"88.99.104.178\" }), String(RadonString { value: \"206.166.251.188\" }), String(RadonString { value: \"178.63.62.52\" }), String(RadonString { value: \"136.244.77.174\" }), String(RadonString { value: \"116.202.48.46\" }), String(RadonString { value: \"209.97.138.224\" }), String(RadonString { value: \"95.217.181.71\" }), String(RadonString { value: \"49.12.73.196\" }), String(RadonString { value: \"213.239.207.48\" }), String(RadonString { value: \"206.189.63.220\" }), String(RadonString { value: \"116.202.48.46\" }), String(RadonString { value: \"116.203.46.135\" }), String(RadonString { value: \"195.201.63.43\" }), String(RadonString { value: \"138.201.62.73\" }), String(RadonString { value: \"88.99.218.237\" }), String(RadonString { value: \"116.202.131.22\" }), String(RadonString { value: \"46.4.75.56\" }), String(RadonString { value: \"88.99.104.178\" }), String(RadonString { value: \"168.119.58.62\" }), String(RadonString { value: \"78.46.84.92\" }), String(RadonString { value: \"159.69.74.217\" }), String(RadonString { value: \"213.133.101.58\" }), String(RadonString { value: \"178.63.67.40\" }), String(RadonString { value: \"78.46.86.104\" }), String(RadonString { value: \"207.154.254.153\" }), String(RadonString { value: \"116.202.235.36\" }), String(RadonString { value: \"168.119.4.50\" }), String(RadonString { value: \"167.99.243.125\" }), String(RadonString { value: \"167.172.39.205\" }), String(RadonString { value: \"195.201.206.107\" })], is_homogeneous: true }, max_count: 3 }") })
```

And this would be the tally using the logic from WIP-0018:

```
Tally: RadonTypes::RadonError(UnhandledInterceptV2 { inner: None })
```

And by executing the tally stage locally it is possible to recover the `inner` error:

```
Local tally: RadonTypes::RadonError(UnhandledInterceptV2 { inner: Some(ModeTie { values: RadonArray { value: [String(RadonString { value: "207.154.254.153" }), String(RadonString { value: "46.4.75.56" }), String(RadonString { value: "116.203.46.135" }), String(RadonString { value: "88.99.104.178" }), String(RadonString { value: "176.9.66.252" }), String(RadonString { value: "136.243.64.186" }), String(RadonString { value: "138.201.132.40" }), String(RadonString { value: "168.119.4.50" }), String(RadonString { value: "63.250.47.141" }), String(RadonString { value: "195.201.63.43" }), String(RadonString { value: "159.69.75.174" }), String(RadonString { value: "159.89.7.193" }), String(RadonString { value: "116.202.131.22" }), String(RadonString { value: "148.251.128.26" }), String(RadonString { value: "85.114.132.62" }), String(RadonString { value: "149.28.82.190" }), String(RadonString { value: "78.46.84.92" }), String(RadonString { value: "176.9.3.164" }), String(RadonString { value: "167.172.39.205" }), String(RadonString { value: "116.203.225.48" }), String(RadonString { value: "116.202.48.46" }), String(RadonString { value: "138.201.46.114" }), String(RadonString { value: "168.119.119.170" }), String(RadonString { value: "148.251.128.18" }), String(RadonString { value: "95.217.181.71" }), String(RadonString { value: "95.111.244.179" }), String(RadonString { value: "148.251.152.217" }), String(RadonString { value: "136.244.77.174" }), String(RadonString { value: "46.4.217.205" }), String(RadonString { value: "78.46.40.149" }), String(RadonString { value: "178.63.62.52" }), String(RadonString { value: "168.119.172.143" }), String(RadonString { value: "68.183.39.75" }), String(RadonString { value: "207.154.238.90" }), String(RadonString { value: "88.198.7.148" }), String(RadonString { value: "149.7.16.186" }), String(RadonString { value: "142.93.250.57" }), String(RadonString { value: "168.119.119.170" }), String(RadonString { value: "178.63.67.40" }), String(RadonString { value: "49.12.73.196" }), String(RadonString { value: "88.99.218.237" }), String(RadonString { value: "94.130.223.45" }), String(RadonString { value: "136.243.93.244" }), String(RadonString { value: "138.201.83.20" }), String(RadonString { value: "168.119.15.50" }), String(RadonString { value: "195.201.206.107" }), String(RadonString { value: "195.201.61.247" }), String(RadonString { value: "138.201.132.240" }), String(RadonString { value: "49.12.73.196" }), String(RadonString { value: "159.69.75.174" }), String(RadonString { value: "206.189.63.220" }), String(RadonString { value: "138.201.66.13" }), String(RadonString { value: "178.63.70.230" }), String(RadonString { value: "159.89.7.193" }), String(RadonString { value: "168.119.119.170" }), String(RadonString { value: "135.181.56.50" }), String(RadonString { value: "167.172.39.205" }), String(RadonString { value: "78.46.86.104" }), String(RadonString { value: "149.28.82.190" }), String(RadonString { value: "88.99.68.109" }), String(RadonString { value: "138.201.65.62" }), String(RadonString { value: "176.9.120.66" }), String(RadonString { value: "116.202.48.46" }), String(RadonString { value: "138.201.132.40" }), String(RadonString { value: "116.202.235.36" }), String(RadonString { value: "136.244.77.174" }), String(RadonString { value: "168.119.58.62" }), String(RadonString { value: "159.69.74.217" }), String(RadonString { value: "167.99.243.125" }), String(RadonString { value: "46.4.75.56" }), String(RadonString { value: "178.63.54.115" }), String(RadonString { value: "88.99.104.178" }), String(RadonString { value: "168.119.172.143" }), String(RadonString { value: "88.99.68.109" }), String(RadonString { value: "136.243.94.171" }), String(RadonString { value: "209.97.138.224" }), String(RadonString { value: "206.166.251.188" }), String(RadonString { value: "159.69.72.188" }), String(RadonString { value: "213.239.234.132" }), String(RadonString { value: "85.114.132.71" }), String(RadonString { value: "135.181.60.189" }), String(RadonString { value: "46.4.88.189" }), String(RadonString { value: "138.201.62.73" }), String(RadonString { value: "135.181.60.183" }), String(RadonString { value: "188.40.94.103" }), String(RadonString { value: "159.89.7.193" }), String(RadonString { value: "135.181.60.181" }), String(RadonString { value: "213.239.207.48" }), String(RadonString { value: "213.133.101.58" }), String(RadonString { value: "167.99.246.148" }), String(RadonString { value: "116.202.218.95" }), String(RadonString { value: "136.243.109.250" }), String(RadonString { value: "116.203.46.135" }), String(RadonString { value: "168.119.172.143" }), String(RadonString { value: "199.192.30.223" }), String(RadonString { value: "78.46.91.66" }), String(RadonString { value: "136.244.77.174" }), String(RadonString { value: "88.99.68.56" })], is_homogeneous: true }, max_count: 3 }) })
```

### Created in retrieval/aggregation

The other kind of `UnhandledIntercept` can be seen in data request `463bb51d32942fb48edb88684a47440d8165bc97a1912acfab220f6de0e87ff5`. In this request a majority of witnesses report an error, and the result of the request is a mode of the different errors, which happens to be an `UnhandledIntercept` error. This is the tally using the old validation rules:

```
Tally: RadonTypes::RadonError(UnhandledIntercept { inner: None, message: Some("inner: Decode { from: \"cbor::value::Value\", to: \"RadonArray\" }") })
```

And this would be the tally using the logic from WIP-0018:

```
Tally: RadonTypes::RadonError(UnhandledInterceptV2 { inner: None })
```

However, in this case it would be impossible to recover the original error, because the reveals do not contain the error message anymore:

```
Tally: RadonTypes::RadonError(UnhandledInterceptV2 { inner: None })
```

Therefore, I think we should allow using the old `UnhandledIntercept` with extra message field when used as a reveal, and only remove the message in the tally stage. This PR currently implements that, for testing. With that logic the tally would still be the same `(inner: None)`, but the reveals would contain some potentially helpful error message (same as before WIP0018).

File: data_structures/src/mainnet_validations.rs [hunk 1]
```diff
@@ -90,7 +90,7 @@ impl TapiEngine {
                         bit_counter.votes += 1;
                     }
                     if (epoch_to_update - bit_counter.init) % bit_counter.period == 0 {
-                        if (bit_counter.votes * 100) / bit_counter.period >= 80 {
+                        if (bit_counter.votes * 100) / bit_counter.period >= 10 {
```
Comment (lrubiorod) [inline data_structures/src/mainnet_validations.rs:93]: Dont forget to remove this after testing

Comment (tmpolaczyk) [reply data_structures/src/mainnet_validations.rs]: If all the tests pass when that line changes, it means that that line is not tested :eyes:

File: validations/src/tests/mod.rs [hunk 2]
```diff
@@ -6486,6 +6607,222 @@ fn tally_unserializable_value() {
...
+    );
+    let tally_transaction_no_message = TallyTransaction::new(
+        dr_pointer,
+        tally_value_no_message.clone(),
+        vec![vt0, vt1],
+        slashed,
+        error_witnesses,
+    );
+
+    let mut active_wips = all_wips_active();
```
Comment (lrubiorod) [inline validations/src/tests/mod.rs:6654]: in RNG PR (#2056) I use all_wips_active to have activated also wips in voting process, and in this case I use current_active_wips()

**Review by lrubiorod — APPROVED** [review]

File: src/cli/node/json_rpc_client.rs [hunk 3]
```diff
@@ -921,92 +928,112 @@ pub fn data_request_report(
...
+                // If the result of the data request is `RadError`, we run the tally stage locally
+                // in order to report a better error message.
+                let non_error_min = f64::from(dr_output.min_consensus_percentage) / 100.0;
+                let commits_count = dr_info.commits.len();
+                let report = run_tally_panic_safe(
+                    &reveal_txns,
+                    &dr_output.data_request.tally,
+                    non_error_min,
+                    commits_count,
+                    &witnet_rad::current_active_wips(),
```
Comment (tmpolaczyk) [inline src/cli/node/json_rpc_client.rs:986]: This should include all active wips except WIP-0018, because we want to conserve the error message.

Comment (tmpolaczyk) [reply src/cli/node/json_rpc_client.rs]: That's tricky, so I opened issue #2072, let's implement it in a separate PR.

**Pull request merged**